src/mapstorage/XmlMapStorage.cpp

```diff
@@ -27,11 +35,160 @@
+        return enumToString(enumToType(val), uint(val));
+    }
+
+private:
+    // define a bunch of methods
+    //   static constexpr Type enumToType(RoomAlignEnum) { return Type::RoomAlignEnum; }
+    //   static constexpr Type enumToType(DoorFlagEnum)  { return Type::DoorFlagEnum;  }
+    //   ...
+    // converting an enumeration type to its corresponding Type value,
+    // which can be used as argument in enumToString() and stringToEnum()
+#define DECL(X) \
+    static constexpr Type enumToType(X) { return Type::X; }
+    X_FOREACH_TYPE_ENUM(DECL)
+#undef DECL
+
+    uint stringToEnum(Type type, const QStringView str, bool &fail) const;
+    const QString &enumToString(Type type, uint val) const;
+
+    std::vector<std::vector<QString>> enumToStrings;
+    std::vector<QHash<QStringView, uint>> stringToEnums;
+    QString empty;
+};
+
+XmlMapStorage::Converter::Converter()
+    : enumToStrings{
+#define DECL(X) /* */ {#X},
+#define DECL_(X, ...) {#X},
+        {X_FOREACH_RoomAlignEnum(DECL)},
+        {X_FOREACH_DOOR_FLAG(DECL_)},
+        {X_FOREACH_EXIT_FLAG(DECL_)},
+        {X_FOREACH_RoomLightEnum(DECL)},
+        {X_FOREACH_ROOM_LOAD_FLAG(DECL)},
+        {X_FOREACH_INFOMARK_CLASS(DECL)},
+        {X_FOREACH_INFOMARK_CLASS(DECL)},
+        {X_FOREACH_ROOM_MOB_FLAG(DECL)},
+        {X_FOREACH_RoomPortableEnum(DECL)},
+        {X_FOREACH_RoomRidableEnum(DECL)},
+        {X_FOREACH_RoomSundeathEnum(DECL)},
+        {X_FOREACH_RoomTerrainEnum(DECL)},
+        {X_FOREACH_TYPE_ENUM(DECL)},
+#undef DECL
+#undef DECL_
+    }
+    , stringToEnums{}
+    , empty{}
+{
+    // create the maps string -> enum value for each enum type listed above
+    for (std::vector<QString> &vec : enumToStrings) {
+        stringToEnums.emplace_back();
+        QHash<QStringView, uint> &map = stringToEnums.back();
+        uint val = 0;
+        for (QString &str : vec) {
+            if (str == "UNDEFINED") {
+                str.clear(); // we never save or load the string "UNDEFINED"
+            } else {
+                map[QStringView(str)] = val;
+            }
+            val++;
+        }
+    }
+}
+
+const QString &XmlMapStorage::Converter::enumToString(Type type, uint val) const
+{
+    const uint index = uint(type);
+    if (index < enumToStrings.size() && val < enumToStrings[index].size()) {
+        return enumToStrings[index][val];
+    }
+    qWarning().noquote().nospace()
+        << "Attempt to save an invalid enum type = " << toString(type) << ", value = " << val
+        << ". Either the current map is damaged, or there is a bug";
+    return empty;
+}
+
+uint XmlMapStorage::Converter::stringToEnum(Type type, const QStringView str, bool &fail) const
+{
+    const uint index = uint(type);
+    if (index < stringToEnums.size()) {
+        auto iter = stringToEnums[index].find(str);
+        if (iter != stringToEnums[index].end()) {
+            return iter.value();
```

Comment:
What if we have a key collision between enums? This solution is very generic which is good from a code duplication standpoint but its brittle in its current form. I'm wondering now if a solution like what we have in `AbstractParser-Room.cpp` might be better: https://github.com/MUME/MMapper/blob/master/src/parser/AbstractParser-Room.cpp#L95-L136

Reply:
I don't see how you can have a key collision:

`stringToEnums` first index is the enum **type**, computed by `enumToType()` which is autogenerated using the macro `X_FOREACH_TYPE_ENUM` which assigns a different value to each type.
Attempts to call `enumToType()` with an enum not listed in `X_FOREACH_TYPE_ENUM` will fail at compile time, because the compiler will not find a match among the overloaded methods  `enumToType()`

`stringToEnums` second index is the enum UPPERCASE name, which must be unique within the enum values - otherwise the enum definition itself will not compile. And of course `QHash<QStringView, int>` is required to handle correctly **hash** collisions, just like `std::unordered_map<>` would do.

Reply:
Gotcha, first level of the data structure is the type and that prevents conflicts across types where an UPPERCASE name might have a conflict (i.e. multiple UNDEFINED, etc).

Reply:
exactly